packages/babel-preset-mapbox/index.js

```diff
@@ -25,130 +42,151 @@ module.exports = function preset(context, opts) {
       // don't work without it: https://github.com/babel/babel/issues/7215
-      name: 'babel-plugin-transform-es2015-destructuring'
-    },
-    {
-      name: 'babel-plugin-transform-class-properties'
+      name: '@babel/plugin-transform-destructuring'
     },
     {
-      // useBuiltIns for these plugins means they'll rely on Object.assign
-      // being available, instead of adding their own polyfill.
-      name: 'babel-plugin-transform-object-rest-spread',
-      options: { useBuiltIns: true }
+      // Enable loose mode to use assignment instead of defineProperty
+      // See discussion in https://github.com/facebook/create-react-app/issues/4263
+      name: '@babel/plugin-proposal-class-properties',
+      options: {
+        loose: true
+      }
     },
     {
-      name: 'babel-plugin-transform-react-jsx',
+      // Note that this assumes `Object.assign` is available.
+      name: '@babel/plugin-proposal-object-rest-spread',
       options: { useBuiltIns: true }
     },
     {
-      name: 'babel-plugin-transform-runtime',
+      // Polyfills the runtime needed for async/await and generators
+      name: '@babel/plugin-transform-runtime',
       options: {
         helpers: false,
-        polyfill: false,
         regenerator: true
       }
-    }
-  ];
+    },
+    {
+      name: '@babel/plugin-syntax-dynamic-import'
+    },
 
-  if (isEnvTest) {
-    payload.push(
-      {
-        name: 'babel-preset-env',
-        options: {
-          targets: {
-            node: 'current'
-          }
+    {
+      name: 'babel-plugin-transform-dynamic-import',
+      conditional: [isEnvTest]
+    },
+    {
+      // Transpiles generator functions to babel regenerator
+      name: '@babel/plugin-transform-regenerator',
+      conditional: [
+        !isEnvTest && {
+          // Async functions are converted to generators by @babel/preset-env
+          async: false
         }
-      },
-      // Compiles import() to a deferred require()
-      {
-        name: 'babel-plugin-dynamic-import-node'
-      }
-    );
-  } else {
-    const presetEnv = {
-      name: 'babel-preset-env',
-      options: {
-        // Do not transform modules to CJS
-        modules: false,
-        useBuiltIns: true
-      }
-    };
-    if (process.env.BROWSERSLIST) {
-      presetEnv.options.targets = {
-        browsers: process.env.BROWSERSLIST
-      };
+      ]
+    },
+    {
+      name: 'babel-plugin-transform-react-remove-prop-types',
+      conditional: [
+        isEnvProduction && {
+          removeImport: true
+        }
+      ]
     }
+  ]
+    .filter(entity => filterConditionals(entity, env))
+    .map(entity => toBabelFormat(entity, opts));
 
-    payload.push(
-      presetEnv,
-      // Adds syntax support for import()
-      { name: 'babel-plugin-syntax-dynamic-import' }
-    );
+  return {
+    plugins,
+    presets
+  };
+}
+
+function filterConditionals(entity, env) {
```

Comment:
🤔 So we end up with two data structures — pre-babel and babel — instead of one. Do you think that this additional layer of abstraction justifies the complexity (instead of using if-then statements to only construct the single babel data structure)?

Reply:
Hmm, a valid point. I ended up creating the `pre-babel` data structure, because of my discontent with `babel` data structure. The real benefit of the abstraction is that it lets me see all the different environment option for a particular preset/plugin. We can achieve that with the regular data structure, do you think it would be a better idea?

Reply:
Up to you, sir.

Reply:
I am going let this stick for now, with an added comment explaining about the data structure.